bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/LivisiWebSocket.java

```diff
@@ -0,0 +1,179 @@
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketConnect;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
+import org.eclipse.jetty.websocket.api.annotations.WebSocket;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.openhab.binding.livisismarthome.internal.listener.EventListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link LivisiWebSocket} implements the websocket for receiving constant updates
+ * from the LIVISI SmartHome web service.
+ *
+ * @author Oliver Kuhl - Initial contribution
+ * @author Sven Strohschein - Renamed from Innogy to Livisi
+ */
+@NonNullByDefault
+@WebSocket
+public class LivisiWebSocket {
+
+    private final Logger logger = LoggerFactory.getLogger(LivisiWebSocket.class);
+
+    private final HttpClient httpClient;
+    private final EventListener eventListener;
+    private final URI webSocketURI;
+    private final int maxIdleTimeout;
+
+    private WebSocketClient client;
+    private @Nullable Session session;
+    private boolean closing;
+
+    /**
+     * Constructs the {@link LivisiWebSocket}.
+     *
+     * @param eventListener the responsible
+     *            {@link org.openhab.binding.livisismarthome.internal.handler.LivisiBridgeHandler}
+     * @param webSocketURI the {@link URI} of the websocket endpoint
+     * @param maxIdleTimeout max idle timeout
+     */
+    public LivisiWebSocket(HttpClient httpClient, EventListener eventListener, URI webSocketURI, int maxIdleTimeout) {
+        this.httpClient = httpClient;
+        this.eventListener = eventListener;
+        this.webSocketURI = webSocketURI;
+        this.maxIdleTimeout = maxIdleTimeout;
+        this.client = createWebSocketClient();
+    }
+
+    /**
+     * Starts the {@link LivisiWebSocket}.
+     */
+    public synchronized void start() throws Exception {
+        if (client.isStopped()) {
+            client = createWebSocketClient();
+            startWebSocketClient(client);
+        }
+
+        session = connectWebSocket(session);
+    }
+
+    private Session connectWebSocket(@Nullable Session session) throws Exception {
+        if (session != null) {
+            session.close();
+        }
+
+        logger.debug("Connecting to LIVISI SmartHome webSocket...");
+        return client.connect(this, webSocketURI).get();
+    }
+
+    /**
+     * Stops the {@link LivisiWebSocket}.
+     */
+    public synchronized void stop() {
+        this.closing = true;
+        if (isRunning()) {
+            closeSession(session);
+        } else {
+            logger.trace("Stopping websocket ignored - was not running.");
+        }
+        session = null;
+        stopWebSocketClient(client);
+        client = createWebSocketClient();
+    }
+
+    private void closeSession(@Nullable Session session) {
```

Comment:
That is to avoid non-null messages. When I can't get rid of the messages when I write

`if(this.session != null) {
this.session.close(); 
}`

because a local variable (or parameter) is required. I think the reason is, that the non-null-check means that this.session could get changed concurrently to NULL (between line 1 and line 2 of the example).

I'm changing line 83-85 to use the same closeSession method.

Reply:
Correct. The usual way to mitigate that would be something like this, i.e. storing the instance variable in a local variable to avoid the potential concurrency problem:
```
        Session session = this.session;
        if (session != null) {
            session.close();
        }
        this.session = null;
```

Reply:
yes, I did that before, but I thought it's confusing to have 2 different variables with the name in the same context/block and a slightly different meaning... And I think it's a good practice to have small(er) methods with a clear naming. Therefore I started to extract it to small methods. I applied this approach to various classes. Is that ok or have I to change that again?

Reply:
Well, it's a matter of preference, another way to write it:
```
        Session localSession = session;
        if (localSession != null) {
            localSession.close();
        }
        session = null;
```

I personally prefer distinct `this.` to keep it clear that it's in fact the same object. Sometimes I have seen instance variable `host` being copied to a local variable called `localHost`, which is confusing to me. :)

You don't have to rewrite it, I won't insist, but it's a bad practice (from a clean code perspective) to accept `null` as argument. In general, it would be better to throw `IllegalArgumentException` to fail fast, and make sure that callers avoid calling with arguments that doesn't make sense. A method called `closeSession` is expected to accept a session as argument and close that session. Instead it will accept null and silently ignore this.

Reply:
I keep the private method approach, I think it's less confusing and short methods are a good practice. It's also more explicit that there is a nullable variable. I'm wondering that your approach doesn't require a Nullable annotation, because localSession is acutally Nullable which isn't visible within the method.